From e87b3f4a79175f73b0865ecd87294af9acd51f64 Mon Sep 17 00:00:00 2001
From: Stuart Menefy <stuart.menefy@codasip.com>
Date: Tue, 23 Jan 2024 17:31:16 +0000
Subject: [PATCH 12/17] hw/riscv/hobgoblin: Add SRAM block

The Hobgoblin platform has an SRAM block internal to the FPGA.

Signed-off-by: Stuart Menefy <stuart.menefy@codasip.com>
---
 hw/riscv/hobgoblin.c         | 10 +++++++++-
 include/hw/riscv/hobgoblin.h |  1 +
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/hw/riscv/hobgoblin.c b/hw/riscv/hobgoblin.c
index ef7741368f..b3e8ea2493 100644
--- a/hw/riscv/hobgoblin.c
+++ b/hw/riscv/hobgoblin.c
@@ -57,6 +57,7 @@ static const struct MemmapEntry {
     [HOBGOBLIN_MROM] =     {     0x1000,      0x100 },
     [HOBGOBLIN_BOOT_ROM] = { 0x10000000, 0x00010000 },
     [HOBGOBLIN_BOOT_RAM] = { 0x10400000, 0x00008000 },
+    [HOBGOBLIN_SRAM] =     { 0x20000000, 0x08000000 },
     [HOBGOBLIN_PLIC] =     { 0x40000000,  0x4000000 },
     [HOBGOBLIN_CLINT] =    { 0x60014000,     0xc000 },
     [HOBGOBLIN_UART0] =    { 0x60100000,     0x1020 },
@@ -93,6 +94,7 @@ static void hobgoblin_board_init(MachineState *machine)
     MemoryRegion *mask_rom = g_new(MemoryRegion, 1);
     MemoryRegion *boot_rom = g_new(MemoryRegion, 1);
     MemoryRegion *boot_ram = g_new(MemoryRegion, 1);
+    MemoryRegion *sram = g_new(MemoryRegion, 1);
     int i;
     hwaddr start_addr;
     uint64_t kernel_entry = 0;
@@ -115,7 +117,7 @@ static void hobgoblin_board_init(MachineState *machine)
     sysbus_realize(SYS_BUS_DEVICE(&s->soc), &error_fatal);
 
     /* register system main memory (actual RAM) */
-    memory_region_init_ram(main_mem, NULL, "riscv.hobgoblin.ram",
+    memory_region_init_ram(main_mem, NULL, "riscv.hobgoblin.dram",
                            memmap[HOBGOBLIN_DRAM].size, &error_fatal);
     memory_region_add_subregion(system_memory, memmap[HOBGOBLIN_DRAM].base,
                                 main_mem);
@@ -138,6 +140,12 @@ static void hobgoblin_board_init(MachineState *machine)
     memory_region_add_subregion(system_memory, memmap[HOBGOBLIN_BOOT_RAM].base,
                                 boot_ram);
 
+    /* SRAM */
+    memory_region_init_ram(sram, NULL, "riscv.hobgoblin.ram",
+                           memmap[HOBGOBLIN_SRAM].size, &error_fatal);
+    memory_region_add_subregion(system_memory, memmap[HOBGOBLIN_SRAM].base,
+                                sram);
+
     /* MMIO */
     s->plic = sifive_plic_create(
         memmap[HOBGOBLIN_PLIC].base,
diff --git a/include/hw/riscv/hobgoblin.h b/include/hw/riscv/hobgoblin.h
index 3c57c3cebb..e4b94c4277 100644
--- a/include/hw/riscv/hobgoblin.h
+++ b/include/hw/riscv/hobgoblin.h
@@ -43,6 +43,7 @@ enum {
     HOBGOBLIN_MROM,
     HOBGOBLIN_BOOT_ROM,
     HOBGOBLIN_BOOT_RAM,
+    HOBGOBLIN_SRAM,
     HOBGOBLIN_PLIC,
     HOBGOBLIN_CLINT,
     HOBGOBLIN_UART0,
-- 
2.39.3

